Smspec cmp + correct padding of ecl_kw string keywords

joakim-hove opened the pull request:
**Task**
This was mostly a wild goose chase trying to find a problem with the summary files generated by OPM. The third commit turned out to be the crucial one. The sorting is also valuable - although not as essential.

**Approach**
1. Make sure every item in ecl_kw of type `ECL_STRING` is space padded to correct length.
2. Implemented function to compare smspec node instances.

**Pre un-WIP checklist**
- [x] Statoil tests pass locally

pgdr commented on lib/ecl/ecl_kw.c:
```diff
@@ -866,11 +868,18 @@ void ecl_kw_iset_string_ptr( ecl_kw_type * ecl_kw, int index, const char * s) {
+      ecl_string[i] = ' ';
+
+    ecl_string[type_len] = '\0';
+  }
```
I think the following is more readable:

```c
for (int i = 0; i < type_len; ++i) {
  ecl_string[i] = (i < input_len) ? s[i] : ' ';
}
```
And also, you don't need the `{` and the `}` guards here.

pgdr commented on lib/ecl/smspec_node.c:
```diff
@@ -980,3 +980,208 @@ void smspec_node_fprintf( const smspec_node_type * smspec_node , FILE * stream)
+  if (node2_early) {
+    if (!node1_early)
+      return 1;
+  }
```
I think the following is more readable:

```c
  if (node1_early && !node2_early)
    return -1;
  if (!node1_early && node2_early)
    return 1;
```

pgdr commented on lib/ecl/tests/ecl_smspec_node.c:
```diff
@@ -0,0 +1,127 @@
+  smspec_node_free( well_node1_2 );
+  smspec_node_free( well_node2_2 );
+  //smspec_node_free( well_node_dummy );
+}
```
There are a couple of out-commented lines here, should they be removed? (Perhaps they are removed in a later commit.)

pgdr commented on python/python/ecl/ecl/ecl_smspec_node.py:
```diff
@@ -41,12 +41,47 @@ class EclSMSPECNode(BaseCClass):
+            return True
+        else:
+            return False
+
```
this should be
```python
    return self.cmp(other) < 0
```

pgdr commented on python/python/ecl/ecl/ecl_smspec_node.py:
```diff
@@ -41,12 +41,47 @@ class EclSMSPECNode(BaseCClass):
+        if cmp > 0:
+            return True
+        else:
+            return False
```
see above, but with  `>`

pgdr commented on python/python/ecl/ecl/ecl_smspec_node.py:
```diff
@@ -41,12 +41,47 @@ class EclSMSPECNode(BaseCClass):
+        if cmp == 0:
+            return True
+        else:
+            return False
```
```python
return self.cmp(other) == 0
```

pgdr commented on python/tests/ecl/test_sum.py:
```diff
@@ -70,19 +70,25 @@ def test_identify_var_type(self):
+        self.assertEqual( node3.getNum( ) , 1 )
+        self.assertTrue( node3.isTotal( ))
+
+        self.assertTrue( node1 < node3 )
```
I would prefer `self.assertLess(node1, node3)`

pgdr commented on python/tests/ecl/test_sum.py:
```diff
@@ -70,19 +70,25 @@ def test_identify_var_type(self):
+        self.assertTrue( node1 < node3 )
+        self.assertTrue( node2 > node3 )
+        self.assertTrue( node1 == node1 )
+        self.assertFalse( node1 == node2 )
```
`assertEqual` and `assertNotEqual`.

pgdr commented on lib/ecl/smspec_node.c:
```diff
@@ -74,32 +74,14 @@ struct smspec_node_struct {
+  if (cmp == 0)
+    return true;
+  else
+    return false;
```
Why not
```c
return smspec_node_cmp(node1, node2) == 0;
```
?

pgdr commented on lib/ecl/tests/ecl_smspec.c:
```diff
@@ -23,6 +23,17 @@
 
+void test_sort( ecl_smspec_type * smspec )
+{
+  ecl_smspec_sort( smspec );
```
This is my paranoia speaking, but could you add
```c
test_assert_true(ecl_smspec_num_nodes(smspec) == 10);
```
for `10` being the correct size of `smspec`?

joakim-hove replied:
I quite frankly do not know the number of nodes, but I have added a check verifying that is not changed by the sort.

joakim-hove commented:
Thank you for the review - of course I will consider your comments.

joakim-hove commented:
OK - I have addressed all the comments except: https://github.com/Statoil/libecl/pull/146#pullrequestreview-53622336 - where I really thought my initial version was more readable. Will merge when/if the rebased tests are green.

pgdr commented:
Fair enough, thanks for addressing!

Please merge at will.

pgdr commented:
Thar she blows!
